airflow/models/dag.py

```diff
@@ -1269,16 +1269,17 @@ def get_task_instances_before(
         ``base_date``, or more if there are manual task runs between the
         requested period, which does not count toward ``num``.
         """
-        min_date = (
-            session.query(DagRun)
+        min_date: Optional[datetime] = (
+            session.query(DagRun.execution_date)
             .filter(
                 DagRun.dag_id == self.dag_id,
                 DagRun.execution_date <= base_date,
                 DagRun.run_type != DagRunType.MANUAL,
             )
             .order_by(DagRun.execution_date.desc())
             .offset(num)
-            .first()
+            .limit(1)
+            .scalar()
```

Comment:
```suggestion
            .scalar()
```
I think using only `.scalar()` is enough.

Reply:
Docs say `scalar()` fails if there’s more than one row

https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.scalar

Reply:
Curious why you went with `limit` instead of `first` ?

https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.first

Reply:
I think `first()` would require unpacking the returning tuple manually? This feels slightly easier to read for me than

```python
min_date_row = session.query(...)
if min_date_row is None:
    ...
min_date = min_date_row[0]
```